### ISSUE TITLE
Update documentation to require python 2.7

### ISSUE THREAD

**Issue by henn** [CCI-MOC/hil#700]
2.6 is ancient, and AFAIK we have no CentOS 6 machines any more.

Any objections?

It would be good to start going to python 3, but let's go one step at a time.

**Comment by zenhack** [comment]
No objections here.

For posterity: the conversation in #680 provides some context; basically @kylehogan submitted code that requires 2.7+ and I mentioned that we should be explicit about what to support.

re: the python 3 thing, another intermediate step would just be writing portable code, which isn't actually that hard. The `six` library smooths over most of the pain points. And we can add 3.x to our travis matrix; I believe there's even a "fail ok" option we could use in the interim, while stuff isn't fully portable still. I don't really care about python 3 myself, but I'll happily review PRs to improve portability.

**Comment by shwsun** [comment]
I agree with removing obsolete things and improving portability

**Comment by henn** [comment]
The [README](https://github.com/CCI-MOC/hil/blob/master/README.rst) says that we require python 2.7, so that portion is solved, but I like the thoughts from @zenhack for slowly transitioning to python 3, so let's make this issue about that.

**Comment by xuhang57** [comment]
`six` ? 

python 2to3?

**Comment by zenhack** [comment]
@xuhang57, the trouble with 2to3 is that it leaves your code in a state where it *only* works with python 3, which is not a leap I'm willing to make just yet; there are a lot of systems out there where python 2 is still the default.

re six: what is your question exactly?

Like I mentioned above, step 1 is probably to add python 3 to the build matrix with a "fail ok" setting, so we can see what's working and what's not. From there we can start fixing stuff.

**Comment by xuhang57** [comment]
Oh, sorry, I was just writing all packages I know that might be helpful for translating python2 to 3.

And what you said about step 1 makes sense to me. But should we create a new branch or even a repo to get started testing? I feel like it makes sense to give it a try and see what we need for the process as early as we could.

**Comment by zenhack** [comment]
@xuhang57, I think someone should submit a PR that adds an item to our build matrix to run the tests against python 3, and flags them allowed to fail. We're doing this for obmd for the (unreleased) master branch of Go, see:

https://github.com/CCI-MOC/obmd/blob/master/.travis.yml#L9-L10

We'll review and merge that PR, which will no doubt have a bunch of failures on python 3. From there we can start submitting PRs fixing the various tests.

**Comment by zenhack** [comment]
Just pushed this to a branch:

https://github.com/zenhack/hil/commit/80b536b6bb83be02c5b169d6aeaf304d27b1d39b

If travis seems to do the right thing with that, I'll submit a pr.

**Comment by xuhang57** [comment]
`virtualenv` seems failed: https://travis-ci.org/zenhack/hil/jobs/344563309

I feel like that is because it should use `virtualenv3` rather than `virtaulenv` which is using the Python 2

**Comment by shwsun** [comment]
It looks more like some ubuntu package installation failure. I think we can just rerun the travis test to see how it goes.

I agree with @zenhack  that the python3 migration should be a long term job. We just don't have enough developers right now.

**Comment by zenhack** [comment]
Yeah, that looks like some kind of transient network failure. I just restarted the build, we'll see how it goes.

**Comment by zenhack** [comment]
Failed again. Related:

https://github.com/travis-ci/travis-ci/issues/9112

But I think this may not be the only issue. I just pushed another commit that tries using 3.4; see the comment there. We'll see if it does any better.

**Comment by xuhang57** [comment]
Should we use this issue to keep track of the Python 3 support? We just started

**Comment by zenhack** [comment]
Works for me.

Quoting Lucas H. Xu (2018-03-06 14:08:56)

>    Should we use this issue to keep track of the Python 3 support? We just
>    started

**Comment by zenhack** [comment]
So, we should probably come up with a triage strategy for this; we're running the CI against python 3, but that as far as I can see nobody (including myself) is really paying attention to it.

**Comment by xuhang57** [comment]
Maybe having a doc/list to keep track of the problems we have known and whoever fixes a/a few problem(s), also updates the doc/list?

**Comment by zenhack** [comment]
It's a bit hard to just enumerate a list when what we're seeing is a test suite that fails up-front. Right now we're getting an error early on. Ideally we'd be able to impose a rule of no *new* test failures. Might be hard to automate that.

**Comment by zenhack** [comment]
Bringing this up again, I'd like to start tackling this. We've got some time, but I'd like to switch over before python 2.7 goes EOL in 2020. This isn't as hard a deadline as it seems; as long as it's the system default on supported versions of RHEL it'll still get security patches, and it hasn't gotten much else in a *looong* time, so frankly the PSF dropping support is more symbolic than anything else. That said, I think it's worth it to start the porting work. Once hil-vpn is ready to deploy, I think I'll shift my efforts over to this. I may do some bits in parallel when the review pipeline gets deep.
